src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -1084,30 +1084,32 @@ const Type* PhiNode::Value(PhaseGVN* phase) const {
     return Type::TOP;
 
   // Check for trip-counted loop.  If so, be smarter.
-  CountedLoopNode* l = r->is_CountedLoop() ? r->as_CountedLoop() : NULL;
+  BaseCountedLoopNode* l = r->is_BaseCountedLoop() ? r->as_BaseCountedLoop() : NULL;
   if (l && ((const Node*)l->phi() == this)) { // Trip counted loop!
     // protect against init_trip() or limit() returning NULL
     if (l->can_be_counted_loop(phase)) {
-      const Node *init   = l->init_trip();
-      const Node *limit  = l->limit();
+      const Node* init = l->init_trip();
+      const Node* limit = l->limit();
       const Node* stride = l->stride();
       if (init != NULL && limit != NULL && stride != NULL) {
-        const TypeInt* lo = phase->type(init)->isa_int();
-        const TypeInt* hi = phase->type(limit)->isa_int();
-        const TypeInt* stride_t = phase->type(stride)->isa_int();
+        const TypeInteger* lo = phase->type(init)->isa_integer(l->bt());
+        const TypeInteger* hi = phase->type(limit)->isa_integer(l->bt());
+        const TypeInteger* stride_t = phase->type(stride)->isa_integer(l->bt());
```

Comment:
Do we have assert somewhere which checks that all loop's values (init, limit, stride) have the same type?

Reply:
Thanks for reviewing it.

> Do we have assert somewhere which checks that all loop's values (init, limit, stride) have the same type?

isa_integer() is passed l->bt() which is T_INT for a CountedLoop and T_LONG for LongCountedLoop. This causes isa_integer() to either call isa_int() or isa_long(). So yes, they should all be the same type and the type that matches the loop's type.